cores/arduino/stm32/twi.c

```diff
@@ -534,19 +541,56 @@ void HAL_I2C_AddrCallback(I2C_HandleTypeDef *hi2c, uint8_t TransferDirection, ui
   */
 void HAL_I2C_ListenCpltCallback(I2C_HandleTypeDef *hi2c)
 {
-  uint8_t nbData = 0;
   i2c_t *obj = get_i2c_obj(hi2c);
 
+  /*  Previous master transaction now ended, so inform upper layer if needed
+   *  then prepare for listeing to next request */
   if((obj->i2c_onSlaveReceive != NULL) &&
     (obj->slaveMode == SLAVE_MODE_RECEIVE)) {
-    nbData = I2C_TXRX_BUFFER_SIZE - obj->handle.XferSize;
-    if(nbData != 0) {
-      obj->i2c_onSlaveReceive(obj->i2cTxRxBuffer, nbData);
+    if(obj->slaveRxNbData != 0) {
+      obj->i2c_onSlaveReceive(obj->i2cTxRxBuffer, obj->slaveRxNbData);
     }
   }
+  obj->slaveMode = SLAVE_MODE_LISTEN;
+  obj->slaveRxNbData = 0;
   HAL_I2C_EnableListen_IT(hi2c);
 }
 
+/**
+  * @brief Slave RX complete callback
+  * @param  hi2c Pointer to a I2C_HandleTypeDef structure that contains
+  *                the configuration information for the specified I2C.
+  * @retval None
+  */
+void HAL_I2C_SlaveRxCpltCallback(I2C_HandleTypeDef *hi2c)
+{
+  i2c_t *obj = get_i2c_obj(hi2c);
+  /* One more byte was received, store it then prepare next */
+  if(obj->slaveRxNbData < I2C_TXRX_BUFFER_SIZE) {
+    obj->slaveRxNbData++;
+  } else {
+    printf("ERROR: I2C Slave RX overflow\n");
+  }
+  /* Restart interrupt mode for next Byte */
+  if(obj->slaveMode == SLAVE_MODE_RECEIVE) {
+      HAL_I2C_Slave_Sequential_Receive_IT(hi2c, &(obj->i2cTxRxBuffer[obj->slaveRxNbData]),
+                                          1, I2C_NEXT_FRAME);
+  }
+}
+
+/**
+  * @brief Slave TX complete callback
+  * @param  hi2c Pointer to a I2C_HandleTypeDef structure that contains
+  *                the configuration information for the specified I2C.
+  * @retval None
+  */
+void HAL_I2C_SlaveTxCpltCallback(I2C_HandleTypeDef *hi2c)
+{
+  i2c_t *obj = get_i2c_obj(hi2c);
+  /*  reset transmit buffer size */
```

Comment:
Please, remove one space and use capital letter for first letter.
/* Reset

Reply:
done